File: 0202-happy-number/linkedlist_cycle_approach.cpp
// Time complexity: O(logn) as number of digits in a number = log n, Space complexity: O(1)
class Solution {
public:
    bool isHappy(int n) {
        
        // The numbers form an implicit cycle that could be solved via the 
        // Floyd Warshall algorithm.
        int slow = n;
        int fast = compSqDigSum(n);

        while(fast != 1 && slow != fast) {
            slow = compSqDigSum(slow);
            fast = compSqDigSum(compSqDigSum(fast));
        }

        return fast == 1;

    }
private:
    int compSqDigSum(int num) {
        int sqSum = 0;

        while(num) {
            int dig = num % 10;
            sqSum += dig*dig;
            num = num / 10;
        }

        return sqSum;
    }
};

File: 0202-happy-number/linkedlist_cycle_approach.py
# Time complexity: O(log N), Space complexity: O(1)
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = n
        fast = self.sumOfSquares(n)

        while fast != 1 and slow != fast:
            fast = self.sumOfSquares(self.sumOfSquares(fast))
            slow = self.sumOfSquares(slow)

            if slow == fast:
                break
        
        return fast == 1

 
    def sumOfSquares(self, n:int) -> int:
        tot = 0

        while n:
            dig = n%10
            dig = dig ** 2
            tot += dig
            n = n // 10

        return tot